### PR TITLE
Allow creating references to nested optionals

### PR DESCRIPTION
**Pull request by SupunS**
Closes https://github.com/onflow/cadence/issues/2541

This PR added two fixes:
- Allow creating nested-optional references (`&T)??`.
- Reject taking references as `&(T?)` 

______

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

### CODE REVIEW

File: runtime/interpreter/interpreter_expression.go [hunk 1]
```diff
@@ -1335,24 +1335,14 @@ func (interpreter *Interpreter) VisitReferenceExpression(referenceExpression *as
...
 
-		return NewEphemeralReferenceValue(
-			interpreter,
-			auth,
-			value,
-			typ.Type,
-			locationRange,
-		)
-	}
+func (interpreter *Interpreter) createReference(
```
Comment (SupunS) [inline runtime/interpreter/interpreter_expression.go:1341]: This is the same old logic, but made it recursive.

File: runtime/tests/interpreter/interpreter_test.go [hunk 2]
```diff
@@ -11514,53 +11514,6 @@ func TestInterpretArrayToConstantSized(t *testing.T) {
 	})
 }
 
-func TestInterpretOptionalReference(t *testing.T) {
```
Comment (SupunS) [inline runtime/tests/interpreter/interpreter_test.go:11517]: moved to `reference_test.go`

File: runtime/tests/checker/reference_test.go [hunk 3]
```diff
@@ -3804,3 +3804,37 @@ func TestCheckReferenceRequiredTypeAnnotation(t *testing.T) {
...
+		require.NoError(t, err)
+	})
+
+	t.Run("reference to nested optional", func(t *testing.T) {
+		t.Parallel()
+
+		_, err := ParseAndCheck(t, `
+            fun main() {
+                var dict: {String: Foo?} = {}
+                var ref: &(Foo??) = &dict["foo"] as &(Foo??)
```
Comment (dsainati1) [inline runtime/tests/checker/reference_test.go:3832]: Did we not previously ban references to optionals, and change them to be optional references? IMO the same logic should apply here; taking a reference to a `T??` type should produce a `(&T)??` type.

Comment (SupunS) [reply runtime/tests/checker/reference_test.go:3832]: > Did we not previously ban references to optionals, and change them to be optional references?

It was supposed to, but this path wasn't handled. This is actually the second part of the issue: https://github.com/onflow/cadence/issues/2541#issuecomment-1879211932. That one still need to be handled. I will open a follow-up PR.

Comment (SupunS) [reply runtime/tests/checker/reference_test.go:3848]: Added the second pat to the same PR, as it turns out to be a simple fix: https://github.com/onflow/cadence/pull/3132/commits/25b0726314c4cd02bd100ae7d41cffa36711f3d3

**Review by turbolent — APPROVED** [review]
Nice!

**Pull request merged**